packages/loopring_v3/contracts/iface/ExchangeData.sol

```diff
@@ -30,191 +16,211 @@ import "./ILoopringV3.sol";
     function MIN_AGE_PROTOCOL_FEES_UNTIL_UPDATED() internal pure returns (uint32) { return 1 days; }
-    function GAS_LIMIT_SEND_TOKENS() internal pure returns (uint32) { return 80000; }
+    function MIN_TIME_IN_SHUTDOWN() internal pure returns (uint32) { return 28 days; }
+    function TX_DATA_AVAILABILITY_SIZE() internal pure returns (uint32) { return 104; }
+    function MAX_AGE_DEPOSIT_UNTIL_WITHDRAWABLE_UPPERBOUND() internal pure returns (uint32) { return 14 days; }
+
+
+    struct AccountLeaf
+    {
+        uint24   accountID;
+        address  owner;
+        uint     pubKeyX;
+        uint     pubKeyY;
+        uint32   nonce;
+        uint     walletHash;
+    }
+
+    struct BalanceLeaf
+    {
+        uint16   tokenID;
+        uint96   balance;
+        uint96   index;
+        uint     tradeHistoryRoot;
+    }
+
+    struct MerkleProof
+    {
+        ExchangeData.AccountLeaf accountLeaf;
+        ExchangeData.BalanceLeaf balanceLeaf;
+        uint[36]                 accountMerkleProof;
+        uint[18]                 balanceMerkleProof;
+    }
+
+    struct BlockContext
+    {
+        bytes32 DOMAIN_SEPARATOR;
+    }
 
     // Represents the entire exchange state except the owner of the exchange.
     struct State
     {
         uint    id;
         uint    exchangeCreationTimestamp;
-        address payable operator; // The only address that can submit new blocks.
-        bool    onchainDataAvailability;
+        bool    rollupMode;
+        uint32  maxAgeDepositUntilWithdrawable;
         bytes32 genesisMerkleRoot;
+        bytes32 DOMAIN_SEPARATOR;
 
         ILoopringV3      loopring;
         IBlockVerifier   blockVerifier;
+        IAgentRegistry   agentRegistry;
         IDepositContract depositContract;
 
-        address lrcAddress;
 
-        uint    totalTimeInMaintenanceSeconds;
-        uint    numDowntimeMinutes;
-        uint    downtimeStart;
 
-        address addressWhitelist;
-        uint    accountCreationFeeETH;
-        uint    accountUpdateFeeETH;
-        uint    depositFeeETH;
-        uint    withdrawalFeeETH;
+        // List of all tokens
+        Token[] tokens;
 
-        Token[]     tokens;
-        Account[]   accounts;
-        Deposit[]   deposits;
-        Request[]   depositChain;
-        Request[]   withdrawalChain;
+        // List of all blocks
+        BlockInfo[] blocks;
+
+        // A map from a token to its tokenID + 1
+        mapping (address => uint16) tokenToTokenId;
 
         // The merkle root of the offchain data stored in a Merkle tree. The Merkle tree
         // stores balances for users using an account model.
         bytes32 merkleRoot;
 
-        // The number of blocks that are submitted onchain
-        uint32 numBlocksSubmitted;
+        // A map from an accountID to a tokenID to if the balance is withdrawn
+        mapping (uint24 => mapping (uint16 => bool)) withdrawnInWithdrawMode;
```

Comment:
Already done in #1450 (not for dynamic increase though, the conclusion in #1472  was that it's not really necessary because the cost difference compared to just doing the full height of 32 compared to 24 is low).